### PR TITLE
Updated BOM Generator

### PR DESCRIPTION
**Pull request by Xelynega**
Added part references and whether or not a part is no_load. Fixed bug with part numbers. Changed dependency from python elementTree to lxml because elemenTree cannot move up/down parent/child elements in xml files.

### CODE REVIEW

File: bom/generate_warg_bom.py [hunk 1]
```diff
@@ -1,49 +1,65 @@
...
+
+    #Parse the netlist and store it in an lxml tree
+    tree = ET.parse(netlist_xml)
+
+    #Traverse over every element with a WARG Part Number 
+    for field in tree.findall(".//*[@name='part_num']"):
+        #Check if the element is under a component or a library element
+        if field.getparent().getparent().tag == 'comp':
+            no_load = False
+            #Check if the part is a no_load part
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: It would be preferable to add information about all the available fields instead of adding specifically the no_load field. This way the script won't need maintenance if we need to add a parameter to the components.

File: bom/generate_warg_bom.py [hunk 2]
```diff
@@ -1,49 +1,65 @@
...
-	    dict_writer.writerows(output_dict)
-    print "wrote to {} file.".format(output_filename)
-
-if __name__ == "__main__":
-    main()
+import pip
+
+try:
+    from lxml import etree as ET
+except ImportError:
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: I've never seen this kind of import before. looks a lot cleaner for installation, I like it.

File: bom/generate_warg_bom.py [hunk 3]
```diff
@@ -1,49 +1,65 @@
...
+
+    reader = csv.DictReader(open('component_library.csv', 'rb'))
+
+    #Add the values from the component_library to the output dictionary
+    component_library_dict = []
+    for line in reader:
+        if "part_num" not in line:
+            print "line doesn't have part_num! line: ", line
+        elif line['part_num'] in found_parts:
+            output_dict[found_parts.index(line['part_num'])].update(line)
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: If there is no component in the component_library.csv file, there should be some kind of warning printed out to the user. I should have added this error check initially but forgot to

**Review by scholvat — CHANGES_REQUESTED** [review]
Looks good, definately an improvement. It would be better if the script could fetch all of the fields for a component instead of specifically getting the no_load field.

File: bom/generate_warg_bom.py [hunk 4]
```diff
@@ -1,49 +1,104 @@
...
-	    dict_writer.writerows(output_dict)
-    print "wrote to {} file.".format(output_filename)
-
-if __name__ == "__main__":
-    main()
+import pip
+
+try:
+    from lxml import etree as ET
+except ImportError:
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: Why did you change from xml to lxml? xml comes standard in python, lxml needs to be installed

**Review by scholvat — COMMENTED** [review]
Can you explain why you switched xml libraries, or switch back to the one that is included in python?

**Comment by Xelynega** [comment]
I used lxml because it's the only one I've used before and because I had no idea how to use XPath. Changed it to use pythons built-in library and found out that XPath isn't that bad.

File: bom/generate_warg_bom.py [hunk 5]
```diff
@@ -1,49 +1,101 @@
...
+                        matching_part = check
+
+            if matching_part == {}:      
+                found_parts.append(part.copy())
+                part.update({'quantity':1, 'ref':ref})
+                output_dict.append(part)
+            else:
+                output_dict[found_parts.index(part)]['quantity'] += 1
+                output_dict[found_parts.index(part)]['ref'] += (', ' + ref)
+    reader = csv.DictReader(open('component_library.csv', 'rb'))
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: component_library.csv is not in the same directory as the python script. See the original code I wrote to get the proper filename, and either do something similar to (or exactly) that, or move the python script into the same root directory.

File: bom/generate_warg_bom.py [hunk 6]
```diff
@@ -1,49 +1,101 @@
...
+    found_parts = []
+    output_dict = []
+    keylist = ['ref', 'quantity']
+
+    #Parse the netlist and store it in an lxml tree
+    tree = ET.parse(netlist_xml)
+
+    #Get a list of all the keys for the CSV at the end
+    for part_element in tree.findall(".//*[@name='part_num']...."):
+        print part_element
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: Remove debug print statement

File: bom/generate_warg_bom.py [hunk 7]
```diff
@@ -1,49 +1,101 @@
...
+    for part in output_dict:
+        try:   
+            if part['Distributer #'] != '':
+                continue
+        except:
+            missing_parts.append(part)
+
+    #Give error with missing parts
+    warning_message = "\n\tWARNING: Parts "
+    for p in missing_parts:
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: Don't use p as a variable name

File: bom/generate_warg_bom.py [hunk 8]
```diff
@@ -1,49 +1,101 @@
...
+            print "line doesn't have part_num! line: ", line
+        for part in output_dict:
+            if part['part_num'] == line['part_num']:
+                part.update(line)
+
+    #Check for parts missing from the library
+    missing_parts = []
+    for part in output_dict:
+        try:   
+            if part['Distributer #'] != '':
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: Parts can be in the WARG library and not have distributer part numbers. Maybe add some logic to line 74 to check if there is a matching part number in the component library.

File: bom/generate_warg_bom.py [hunk 9]
```diff
@@ -1,49 +1,101 @@
...
+                continue
+        except:
+            missing_parts.append(part)
+
+    #Give error with missing parts
+    warning_message = "\n\tWARNING: Parts "
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: You can have have a separate print statement for each warning, there isn't a need for this loop, it just complicates the code.

**Review by scholvat — CHANGES_REQUESTED** [review]
I tried running the program for the first time and found some more bugs

File: bom/generate_warg_bom.py [hunk 10]
```diff
@@ -1,49 +1,92 @@
...
+                        matching_part = check
+
+            if matching_part == {}:      
+                found_parts.append(part.copy())
+                part.update({'quantity':1, 'ref':ref})
+                output_dict.append(part)
+            else:
+                output_dict[found_parts.index(part)]['quantity'] += 1
+                output_dict[found_parts.index(part)]['ref'] += (', ' + ref)
+    reader = csv.DictReader(open('../component_library.csv', 'rb'))
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: Still not working on my machine... maybe us the os library like I originally had.

File: bom/generate_warg_bom.py [hunk 11]
```diff
@@ -1,49 +1,92 @@
...
-    #write the csv
-    with open(output_filename, 'wb') as output_file:
-	    dict_writer = csv.DictWriter(output_file, output_dict[0].keys())
-	    dict_writer.writeheader()
-	    dict_writer.writerows(output_dict)
-    print "wrote to {} file.".format(output_filename)
-
-if __name__ == "__main__":
-    main()
+import pip
```
Comment (scholvat) [inline bom/generate_warg_bom.py]: remove pip since I don't think you need it anymore

**Review by scholvat — CHANGES_REQUESTED** [review]

**Review by scholvat — APPROVED** [review]

**Comment by scholvat** [comment]
Looks good, I tested it on both Window and Linux and runs fine.
